File: 0139-word-break/0139-word-break.dart
class Solution {
  bool wordBreak(String s, List<String> wordDict) {
    Set<String> dict = wordDict.toSet();
    for(int i=0;i<s.length;i++){
        String sub = s.substring(0, i);
        if(!dict.contains(sub)) continue;
        for(String word in wordDict){
            String newWord = sub + word;
            if(s == newWord) return true;
            dict.add(newWord);
        }
    }
    return dict.contains(s);
  }
}

File: 0139-word-break/0139-word-break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        vector<bool> dp(n+1);
        dp[0] = true;
        for(int i=0;i<n;i++){
            if(!dp[i]) continue;
            for(int j=i+1;j<=n;j++){
                string word = s.substr(i, j-i);
                if(dict.contains(word)) {
                    dp[j] = true;
                }
            }
        }
        return dp[n];
    }
};